keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -110,31 +110,76 @@ When this app moves to production on containers, the team will need to coordinat
   - Provision the GMSA and gives details to Kubernetes cluster admin.
   - Assign access to a AD security group to control what machines can use this GMSA. The AD security group should include all authorized Kubernetes nodes.
 
-2. A Kubernetes cluster admin (or cluster setup tools or an operator) will install a cluster scoped CRD (of kind GMSACredSpec) whose instances will be used to store GMSA credential spec configuration:
+2. A Kubernetes cluster admin (or cluster setup tools or an operator) will install a cluster scoped CRD (of kind GMSACredentialSpec) whose instances will be used to store GMSA credential spec configuration:
 
 ```
 apiVersion: apiextensions.k8s.io/v1beta1
 kind: CustomResourceDefinition
 metadata:
-  name: gmsacredspecs.auth.k8s.io
+  name: gmsacredentialspecs.windows.k8s.io
 spec:
-  group: auth.k8s.io
+  group: windows.k8s.io
   version: v1alpha1
   names:
-    kind: GMSACredSpec
-    plural: gmsacredspecs
+    kind: GMSACredentialSpec
+    plural: gmsacredentialspecs
   scope: Cluster
+  validation:
+    openAPIV3Schema:
+      properties:
+        credspec:
```

Comment:
I think it will be better to keep this a string field since nothing in Kubernetes code cares about the contents. The schema is not documented anywhere official like MSDN. So it may change in a future version of Windows. I will get rid of the rest of the schema below.

Reply:
Updated `credspec` to be string rather than specify a more stringent schema which may change in future and not officially documented beyond https://raw.githubusercontent.com/Microsoft/Virtualization-Documentation/live/windows-server-container-tools/ServiceAccounts/CredentialSpec.psm1

Reply:
I think the schema should match the schema of the destination field. If the plan is for the destination to be an annotation or string field in the pod, and a string field in the CRI API, then this should be a string field. Otherwise, admission has to know how to transform this structured data into a string

Reply:
edit: I missed that in the OCI spec, credentialSpec is an object, not a string - https://github.com/opencontainers/runtime-spec/blob/eba862dc2470385a233c7507392675cbeadf7353/schema/config-windows.json#L80-L82

how it would be added to the CRI API is an open question though... can it represent arbitrary structure well or would it be plumbed as a string containing json?

Reply:
@liggitt I can go ahead and modify the type for `credspec` to be string. I agree that it will allow things to be consistent (as type string) through the Pod spec field  => kubelet processing => CRI API => CRI implementation. 

The one main problem with declaring credspec as type string is user friendliness - doing a `kubectl describe` will return a string blob rather that YAML formatted output:
```
API Version:  windows.k8s.io/v1alpha1
Credspec:      ActiveDirectoryConfig: GroupManagedServiceAccounts: - Name: WebApplication1Scope: CONTOSO - Name: WebApplication1 Scope: contoso.com CmsPlugins: - ActiveDirectory DomainJoinConfig: DnsName: contoso.com DnsTreeName: contoso.com Guid: 244818ae-87ca-4fcd-92ec-e79e5252348a MachineAccountName: WebApplication1 NetBiosName: CONTOSO Sid: S-1-5-21-2126729477-2524075714-3094792973
Kind:         GMSACredentialSpec
```
@liggitt @michmike Let me know if there is a preference.

Regarding the type of credentialSpec in OCI, note a couple of things:
1. I don't think CRI API needs to be concerned about being able to pass arbitrary object type (at least right now in the context of the GMSA KEP) since the CRI backend like CRI ContainerD will be the component processing what is passed through the CRI API and creating the OCI spec to pass to ContainerD. Since CRI ContainerD today completely creates the spec from scratch and analyzes and converts various pieces of data from CRI API, it should be able to do process the credential spec in a similar fashion: https://github.com/containerd/cri/blob/master/pkg/server/container_create.go#L215-L238. This necessary enhancement is already called out below in the `Changes in CRIContainerD` section.

2. The underlying type in the spec and what is passed down the Windows platform layer (HCS) is expected to be string:
https://github.com/containerd/containerd/blob/master/windows/hcsshim.go#L113